File: obsidian-mcp-ts/src/obsidian/templates.ts
import fs from "fs";
import path from "path";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { getVaultPath } from "./vault.js";
import { relPath } from "./utils.js";

export function getTemplatesFolder(): string {
  const vaultPath = getVaultPath();
  const candidates = [
    path.join(vaultPath, ".templates"),
    path.join(vaultPath, "Templates"),
    path.join(vaultPath, "templates"),
  ];

  for (const folder of candidates) {
    if (fs.existsSync(folder)) return folder;
  }

  const defaultFolder = path.join(vaultPath, "Templates");
  fs.mkdirSync(defaultFolder, { recursive: true });
  return defaultFolder;
}

export function renderTemplate(templateContent: string, variables: Record<string, string>): string {
  let content = templateContent;

  for (const [key, value] of Object.entries(variables)) {
    const escaped = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    content = content.replace(new RegExp(`\\{\\{${escaped}\\}\\}`, "g"), value);
  }

  const now = new Date();
  const dateVars: Record<string, string> = {
    date: now.toISOString().slice(0, 10),
    time: now.toTimeString().slice(0, 8),
    datetime: `${now.toISOString().slice(0, 10)} ${now.toTimeString().slice(0, 8)}`,
    year: String(now.getFullYear()),
    month: String(now.getMonth() + 1).padStart(2, "0"),
    day: String(now.getDate()).padStart(2, "0"),
    weekday: now.toLocaleDateString("en-US", { weekday: "long" }),
    month_name: now.toLocaleDateString("en-US", { month: "long" }),
  };

  for (const [key, value] of Object.entries(dateVars)) {
    const escaped = key.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
    content = content.replace(new RegExp(`\\{\\{${escaped}\\}\\}`, "g"), value);
  }

  return content;
}

export function registerTemplateTools(server: McpServer) {
  server.tool(
    "list_templates",
    "List all available templates in the vault",
    {},
    async () => {
      try {
        const vaultPath = getVaultPath();
        const templatesFolder = getTemplatesFolder();

        const templates: any[] = [];
        for (const file of fs.readdirSync(templatesFolder)) {
          if (!file.endsWith(".md")) continue;
          const filePath = path.join(templatesFolder, file);
          try {
            const content = fs.readFileSync(filePath, "utf-8");
            templates.push({
              name: path.basename(file, ".md"),
              path: relPath(vaultPath, filePath),
              size: content.length,
              preview: content.length > 200 ? content.slice(0, 200) + "..." : content,
            });
          } catch { continue; }
        }

        templates.sort((a, b) => a.name.localeCompare(b.name));
        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ templates_folder: relPath(vaultPath, templatesFolder), total_templates: templates.length, templates }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message, templates: [] }) }] };
      }
    }
  );

  server.tool(
    "create_note_from_template",
    "Create a new note from a template",
    {
      template_name: z.string().describe("Name of the template (without .md extension)"),
      note_name: z.string().describe("Name for the new note (without .md extension)"),
      variables: z.record(z.string()).optional().describe("Optional dictionary of variables to substitute"),
      folder: z.string().optional().describe("Optional folder path to create note in"),
    },
    async ({ template_name, note_name, variables, folder }) => {
      try {
        const vaultPath = getVaultPath();
        const templatesFolder = getTemplatesFolder();
        const templatePath = path.join(templatesFolder, `${template_name}.md`);

        if (!fs.existsSync(templatePath)) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Template not found: ${template_name}` }) }] };
        }

        const templateContent = fs.readFileSync(templatePath, "utf-8");
        const rendered = renderTemplate(templateContent, variables || {});

        let notePath: string;
        if (folder) {
          const noteDir = path.join(vaultPath, folder);
          fs.mkdirSync(noteDir, { recursive: true });
          notePath = path.join(noteDir, `${note_name}.md`);
        } else {
          notePath = path.join(vaultPath, `${note_name}.md`);
        }

        if (fs.existsSync(notePath)) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Note already exists: ${note_name}` }) }] };
        }

        fs.writeFileSync(notePath, rendered, "utf-8");
        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ success: true, template_name, note_name, path: relPath(vaultPath, notePath), size: rendered.length }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "save_template",
    "Save a template to the templates folder",
    {
      template_name: z.string().describe("Name of the template (without .md extension)"),
      content: z.string().describe("Template content with {{variable}} placeholders"),
    },
    async ({ template_name, content }) => {
      try {
        const vaultPath = getVaultPath();
        const templatesFolder = getTemplatesFolder();
        const templatePath = path.join(templatesFolder, `${template_name}.md`);
        fs.writeFileSync(templatePath, content, "utf-8");
        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ success: true, template_name, path: relPath(vaultPath, templatePath), size: content.length }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );

  server.tool(
    "get_template",
    "Get a template's content",
    { template_name: z.string().describe("Name of the template (without .md extension)") },
    async ({ template_name }) => {
      try {
        const templatesFolder = getTemplatesFolder();
        const templatePath = path.join(templatesFolder, `${template_name}.md`);

        if (!fs.existsSync(templatePath)) {
          return { content: [{ type: "text" as const, text: JSON.stringify({ error: `Template not found: ${template_name}` }) }] };
        }

        const content = fs.readFileSync(templatePath, "utf-8");
        return {
          content: [{
            type: "text" as const,
            text: JSON.stringify({ template_name, content, size: content.length, lines: content.split("\n").length }),
          }],
        };
      } catch (e: any) {
        return { content: [{ type: "text" as const, text: JSON.stringify({ error: e.message }) }] };
      }
    }
  );
}

File: obsidian-mcp-py/obsidian/templates.py
"""
Obsidian Templates Management

Provides tools for managing note templates.
"""

from pathlib import Path
from typing import Optional, Dict
from datetime import datetime
import re
from obsidian.vault import get_vault_path, safe_path


def get_templates_folder() -> Path:
    """
    Get the templates folder path.
    Defaults to vault/.templates/ or vault/Templates/
    
    Returns:
        Path to templates folder
    """
    vault_path = get_vault_path()
    
    # Try common template folder names
    template_folders = [
        vault_path / ".templates",
        vault_path / "Templates",
        vault_path / "templates"
    ]
    
    for folder in template_folders:
        if folder.exists():
            return folder
    
    # Create default template folder if none exists
    default_folder = vault_path / "Templates"
    default_folder.mkdir(exist_ok=True)
    return default_folder


def render_template(template_content: str, variables: Dict[str, str]) -> str:
    """
    Render a template with variables.
    
    Supports {{variable}} syntax for variable substitution.
    
    Args:
        template_content: Template content with {{variable}} placeholders
        variables: Dictionary of variable names to values
    
    Returns:
        Rendered template content
    """
    content = template_content
    
    # Replace {{variable}} with values
    for key, value in variables.items():
        pattern = r'\{\{' + re.escape(key) + r'\}\}'
        content = re.sub(pattern, str(value), content)
    
    # Replace date/time variables
    now = datetime.now()
    date_vars = {
        "date": now.strftime("%Y-%m-%d"),
        "time": now.strftime("%H:%M:%S"),
        "datetime": now.strftime("%Y-%m-%d %H:%M:%S"),
        "year": now.strftime("%Y"),
        "month": now.strftime("%m"),
        "day": now.strftime("%d"),
        "weekday": now.strftime("%A"),
        "month_name": now.strftime("%B")
    }
    
    for key, value in date_vars.items():
        pattern = r'\{\{' + re.escape(key) + r'\}\}'
        content = re.sub(pattern, value, content)
    
    return content


def register_template_tools(mcp):
    """
    Register all template-related MCP tools.
    
    Args:
        mcp: FastMCP instance to register tools with
    """
    
    @mcp.tool()
    def list_templates() -> dict:
        """
        List all available templates in the vault.
        
        Returns:
            Dictionary with list of templates
        """
        try:
            templates_folder = get_templates_folder()
            
            templates = []
            for template_file in templates_folder.glob("*.md"):
                try:
                    content = template_file.read_text(encoding="utf-8")
                    templates.append({
                        "name": template_file.stem,
                        "path": str(template_file.relative_to(get_vault_path())),
                        "size": len(content),
                        "preview": content[:200] + "..." if len(content) > 200 else content
                    })
                except Exception:
                    continue
            
            templates.sort(key=lambda x: x["name"])
            
            return {
                "templates_folder": str(templates_folder.relative_to(get_vault_path())),
                "total_templates": len(templates),
                "templates": templates
            }
        except Exception as e:
            return {"error": str(e), "templates": []}
    
    
    @mcp.tool()
    def create_note_from_template(template_name: str, note_name: str, variables: Optional[Dict[str, str]] = None, folder: Optional[str] = None) -> dict:
        """
        Create a new note from a template.
        
        Args:
            template_name: Name of the template (without .md extension)
            note_name: Name for the new note (without .md extension)
            variables: Optional dictionary of variables to substitute in template
            folder: Optional folder path to create note in
        
        Returns:
            Dictionary with creation status
        """
        try:
            vault_path = get_vault_path()
            templates_folder = get_templates_folder()
            
            template_path = templates_folder / f"{template_name}.md"
            
            if not template_path.exists():
                return {"error": f"Template not found: {template_name}"}
            
            template_content = template_path.read_text(encoding="utf-8")
            
            # Render template with variables
            vars_dict = variables or {}
            rendered_content = render_template(template_content, vars_dict)
            
            # Create note
            if folder:
                note_dir = safe_path(vault_path, folder)
                note_dir.mkdir(parents=True, exist_ok=True)
                note_path = note_dir / f"{note_name}.md"
            else:
                note_path = safe_path(vault_path, f"{note_name}.md")
            
            if note_path.exists():
                return {"error": f"Note already exists: {note_name}"}
            
            note_path.write_text(rendered_content, encoding="utf-8")
            
            return {
                "success": True,
                "template_name": template_name,
                "note_name": note_name,
                "path": str(note_path.relative_to(vault_path)),
                "size": len(rendered_content)
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def save_template(template_name: str, content: str) -> dict:
        """
        Save a template to the templates folder.
        
        Args:
            template_name: Name of the template (without .md extension)
            content: Template content with {{variable}} placeholders
        
        Returns:
            Dictionary with save status
        """
        try:
            templates_folder = get_templates_folder()
            template_path = templates_folder / f"{template_name}.md"
            
            template_path.write_text(content, encoding="utf-8")
            
            return {
                "success": True,
                "template_name": template_name,
                "path": str(template_path.relative_to(get_vault_path())),
                "size": len(content)
            }
        except Exception as e:
            return {"error": str(e)}
    
    
    @mcp.tool()
    def get_template(template_name: str) -> dict:
        """
        Get a template's content.
        
        Args:
            template_name: Name of the template (without .md extension)
        
        Returns:
            Dictionary with template content
        """
        try:
            templates_folder = get_templates_folder()
            template_path = templates_folder / f"{template_name}.md"
            
            if not template_path.exists():
                return {"error": f"Template not found: {template_name}"}
            
            content = template_path.read_text(encoding="utf-8")
            
            return {
                "template_name": template_name,
                "content": content,
                "size": len(content),
                "lines": len(content.splitlines())
            }
        except Exception as e:
            return {"error": str(e)}
